### PR TITLE
Make NodeList readonly struct and optimize indexers

### PR DESCRIPTION
**Pull request by lahma**
As `NodeList` is passed around and used a lot on Jint's side, it makes sense to change from regular `struct` to `readonly struct`. Now there will be less defensive copies when methods (like the indexer) are being used. I also checked the indexers and made them use half the IL code to make them perform better.

Here are benchmark results from Jint where the Esprma usage is fraction of the time used running, but we can see positive boost nevertheless.

## Jint.Benchmark.DromaeoBenchmark

| **Diff**|Method|FileName|Mean|Gen 0|Gen 1|Gen 2|Allocated|
|------- |-------|-------|-------:|-------:|-------:|-------:|-------:|
| Old |Run|dromaeo-3d-cube|59.53 ms|1222.2222|222.2222|-|7417.87 KB|
| **New** |	|	| **59.20 ms (-1%)** | **1222.2222 (0%)** | **222.2222 (0%)** | **-** | **7417.87 KB (0%)** |
| Old |Run|dromaeo-core-eval|13.13 ms|78.1250|-|-|377.66 KB|
| **New** |	|	| **13.04 ms (-1%)** | **78.1250 (0%)** | **-** | **-** | **377.66 KB (0%)** |
| Old |Run|dromaeo-object-array|154.16 ms|41500.0000|1500.0000|250.0000|174933.85 KB|
| **New** |	|	| **151.00 ms (-2%)** | **41500.0000 (0%)** | **1500.0000 (0%)** | **250.0000 (0%)** | **174932.15 KB (0%)** |
| Old |Run|droma(...)egexp [21]|717.10 ms|50000.0000|30000.0000|20000.0000|407356.26 KB|
| **New** |	|	| **697.65 ms (-3%)** | **53000.0000 (+6%)** | **33000.0000 (+10%)** | **21000.0000 (+5%)** | **408566.76 KB (0%)** |
| Old |Run|droma(...)tring [21]|769.28 ms|90000.0000|49000.0000|46000.0000|1378209.27 KB|
| **New** |	|	| **771.58 ms (0%)** | **108000.0000 (+20%)** | **67000.0000 (+37%)** | **64000.0000 (+39%)** | **1378167.91 KB (0%)** |

## Jint.Benchmark.StopwatchBenchmark

| **Diff**|Method|N|Mean|Gen 0|Gen 1|Gen 2|Allocated|
|------- |-------|-------|-------:|-------:|-------:|-------:|-------:|
| Old |Jint|1|913.1 ms|12000.0000|-|-|51.61 MB|
| **New** |	|	| **894.3 ms (-2%)** | **12000.0000 (0%)** | **-** | **-** | **51.61 MB (0%)** |

### CODE REVIEW

File: src/Esprima/ExceptionHelper.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+using System;
+
+namespace Esprima
+{
+    internal static class ExceptionHelper
```
[the remark this answers is not in the document]
Comment (lahma) [reply src/Esprima/ExceptionHelper.cs]: My French is almost as bad as my German, but what I gather you are suggesting more implementation specific name. Here I chose the same name as in Jint, which derives from having a name that derives from role, like in BCL they use [ThrowHelper](https://source.dot.net/#System.Private.CoreLib/shared/System/Collections/Generic/List.cs,49). JitHelper might not be helping in the worst case, but of course we are splitting hairs here.

File: src/Esprima/Ast/SequenceExpression.cs [hunk 2]
```diff
@@ -4,14 +4,16 @@ namespace Esprima.Ast
 {
     public class SequenceExpression : Node, Expression
     {
-        public NodeList<Expression> Expressions { get; internal set; }
+        internal NodeList<Expression> _expressions;
```
[the remark this answers is not in the document]
Comment (lahma) [reply src/Esprima/Ast/SequenceExpression.cs]: I've added the suggested method

**Comment by lahma** [comment]
Thank you @atifaziz for the review, good observations. I've incorporated your feedback and it should now follow your recommendations, except I've kept the ExceptionHelper and didn't split another PR for the indexing optimization (I'm lazy yes, but I did create a separate commit though 😉 ).

File: src/Esprima/ExceptionHelper.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (atifaziz) [reply src/Esprima/ExceptionHelper.cs]: > My French is almost as bad as my German

Sorry about that. Actually [_raison d'être_](https://en.wiktionary.org/wiki/raison_d%27être) is actually perfectly valid in English as it's been borrowed from French verbatim.



> I gather you are suggesting more implementation specific name.

You guessed it quite right!

> Here I chose the same name as in Jint, which derives from having a name that derives from role, like in BCL they use [ThrowHelper](https://source.dot.net/#System.Private.CoreLib/shared/System/Collections/Generic/List.cs,49).

It's good to borrow but it's good to question too. `ThrowHelper` sounds dull and uninspiring. But a name like `JitHelper` can only carry so much meaning…

> JitHelper might not be helping in the worst case

…unless you add a `<summary>` comment.

> …but of course we are splitting hairs here.

Which explains why I don't have much hair left. 😄I guess it's down to [tomayto, tomahto](https://en.wiktionary.org/wiki/tomayto,_tomahto).

As I started out by saying, “small nitpick”.

**Comment by atifaziz** [comment]
> I've incorporated your feedback and it should now follow your recommendations

Tremendous and thanks!

> …except I've kept the ExceptionHelper

[Too bad](https://github.com/sebastienros/esprima-dotnet/pull/109#discussion_r307018321) I couldn't convince you otherwise but okay.

> …and didn't split another PR for the indexing optimization (I'm lazy yes, but I did create a separate commit though 😉 ).

Thumbs-up for a separate commit but it made the review longer nonetheless. This was not a complaint but feedback for the future to help speed up reviews. When one is looking at pages of diffs, it's difficult to be lazy as a reviewer because you're building a mental stack only to unstack and re-stack, if you see what I mean. I don't know how @sebastienros does it, frankly. I need to sit next to him once and learn.

Thanks again for all the good and hard work in this PR!

**Pull request merged**